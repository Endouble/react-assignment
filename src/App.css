@import url('https://fonts.googleapis.com/css?family=Oswald');

.app-header {
  background-color: #222;
  padding: 20px;
}

.app-header h1.ui.header {
  font-family: 'Oswald', sans-serif;
  font-size: 2rem;
}

@media (min-width: 768px) { 
    .app-header h1.ui.header {
        font-size: 5rem;
    }
}