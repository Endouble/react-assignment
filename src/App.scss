// Google Fonts
@import url(http://fonts.googleapis.com/css?family=Roboto+Slab|Open+Sans:400italic,700italic,400,700);

// Font Variables
$roboto-slab: 'Roboto Slab', serif;
$open-sans: 'Open Sans', sans-serif;

// Styles
html,
body {
  font-family: $roboto-slab;
}

.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

.resetButton {
  margin: 0px 10px;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}
