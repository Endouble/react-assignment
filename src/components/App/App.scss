@import '../../styles/base.scss';

.App {
  text-align: center;
}

.inputSearch {
  width: 100%;
  padding: 12px 20px;
  margin: 8px 0;
  box-sizing: border-box;
  border: 2px solid $dark-text;
  border-radius: 4px;
}

.resetButton {
  margin: 8px;
  background-color: $yellow;
}
