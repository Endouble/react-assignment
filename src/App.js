import React from 'react';

import Grid from './components/Grid';
import './App.scss';

const App = () => (
    <div className="App">
        <Grid />
    </div>
);

export default App;
