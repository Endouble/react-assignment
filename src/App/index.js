import React, { Component } from 'react';
import logo from './logo.svg';
import Modal from '../Modal';
import ModalContainer from '../ModalContainer';
import BreweryCard from '../BreweryCard';
import Brewery from '../Brewery';
import './App.scss';

/* eslint-env browser */
class App extends Component {
    constructor(props) {
        super(props);

        this.state = {
            breweries: {},
            filterValue: '',
        };

        this.onBreweryClick = this.onBreweryClick.bind(this);
        this.onFilterChange = this.onFilterChange.bind(this);
        this.onFilterSubmit = this.onFilterSubmit.bind(this);
        this.hideModal = this.hideModal.bind(this);
    }

    componentDidMount() {
        /* eslint-disable react/no-did-mount-set-state */

        fetch('/breweries')
            .then(response => response.json())
            .then((arr) => {
                this.setState({
                    breweries: arr.reduce((memo, brewery) => ({
                        ...memo,
                        [brewery.id]: brewery,
                    }), {}),
                });
            });
    }

    onBreweryClick(evt, breweryId) {
        evt.preventDefault();
        this.showModal(breweryId);
    }

    onFilterChange(evt) {
        this.setState({
            filterValue: evt.target.value,
        });
    }

    onFilterSubmit(evt) {
        evt.preventDefault();

        fetch(`/breweries?by_city=${this.state.filterValue}`)
            .then(response => response.json())
            .then((arr) => {
                this.setState({
                    breweries: arr.reduce((memo, brewery) => ({
                        ...memo,
                        [brewery.id]: brewery,
                    }), {}),
                });
            });
    }

    hideModal() {
        this.setState({
            selectedBreweryId: undefined,
        });
    }

    showModal(breweryId) {
        this.setState({
            selectedBreweryId: breweryId,
        });
    }

    renderBreweries() {
        const { breweries } = this.state;

        if (!Object.keys(breweries).length) {
            return null;
        }

        return (
            <div className="App__breweries">
                {Object.values(breweries).map(this.renderBrewery, this)}
            </div>
        );
    }

    renderBrewery(brewery) {
        return (
            <div className="App__brewery-container" key={brewery.id}>
                <BreweryCard
                    onClick={this.onBreweryClick}
                    brewery={brewery}
                />
            </div>
        );
    }

    renderFiler() {
        return (
            <form onSubmit={this.onFilterSubmit} className="App__search-form">
                <label htmlFor="filter">
                    City name:
                    <input id="filter" type="text" value={this.state.filterValue} onChange={this.onFilterChange} />
                </label>
                <input type="submit" value="Submit" />
            </form>
        );
    }

    renderModal() {
        const { selectedBreweryId } = this.state;
        if (!selectedBreweryId) {
            return null;
        }

        const selectedBrewery = this.state.breweries[selectedBreweryId];

        return (
            <ModalContainer>
                <Modal
                    onCloseClick={this.hideModal}
                >
                    <Brewery brewery={selectedBrewery} />
                </Modal>
            </ModalContainer>
        );
    }

    render() {
        return (
            <div className="App">
                <header className="App__header">
                    <img src={logo} className="App__logo" alt="logo" />
                    <h1 className="App__title">Welcome to React</h1>
                </header>
                <div className="App__body">
                    {this.renderFiler()}
                    {this.renderBreweries()}
                </div>
                {this.renderModal()}
            </div>
        );
    }
}

export default App;
