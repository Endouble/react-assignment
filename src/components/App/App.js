import React, { Component } from 'react';
import logo from '../../logo.svg';
import './App.css';
import LoadingComponent from '../LoadingComponent';
import QuoteListComponent from '../QuoteListComponent';
import HeaderComponent from '../HeaderComponent';
import { API_PATH, QUOTES_CHUNK_SIZE } from '../../constants';

class App extends Component {
    constructor() {
        super();
        this.state = {
            loading: false,
            originalFetchedData: [],
            liveData: [],
            query: '',
            isModalOpen: false,
            openedQuote: {},
        };
        this.handleSearch = this.handleSearch.bind(this);
        this.openModal = this.openModal.bind(this);
        this.closeModal = this.closeModal.bind(this);
    }

    componentDidMount() {
        this.setState({ loading: true }); // eslint-disable-line react/no-did-mount-set-state
        fetch(API_PATH + QUOTES_CHUNK_SIZE) // eslint-disable-line no-undef
            .then(response => response.json())
            .then(data => this.setState({ liveData: data, originalFetchedData: data, loading: false }));
    }

    openModal(index) { // Will be used on the card component
        const { liveData } = this.state;
        this.setState({ isModalOpen: true, openedQuote: liveData[index] });
    }

    closeModal() { // Will be used on the card  list component
        this.setState({ isModalOpen: false, openedQuote: {} });
    }

    handleSearch(query) {
        const { originalFetchedData } = this.state;

        // We're searching inside the quote content, not by author for example
        if (query) {
            const filteredData =
                originalFetchedData.filter(item => item.quote.toUpperCase().includes(query.toUpperCase()));
            return this.setState({ liveData: filteredData, query });
        }

        // Cleaning the search
        return this.setState({ query, liveData: originalFetchedData });
    }

    render() {
        const { liveData, loading, query, isModalOpen, openedQuote } = this.state;

        if (loading) return <LoadingComponent />;

        return (
            <div className="App container">
                <HeaderComponent logo={logo} />
                <div className='row container'>
                    <input
                        value={query}
                        placeholder='Search..'
                        onChange={e => this.handleSearch(e.target.value)}
                        className='col-6 inputSearch'
                    />
                    {query && (
                        <button
                            onClick={() => this.handleSearch('')}
                            className='col-4 resetButton'
                        >
                            <span>Reset search</span>
                        </button>
                    )}
                </div>
                <QuoteListComponent
                    items={liveData}
                    isModalOpen={isModalOpen}
                    openedQuote={openedQuote}
                    openModal={this.openModal}
                    closeModal={this.closeModal}
                />
            </div>
        );
    }
}

export default App;
