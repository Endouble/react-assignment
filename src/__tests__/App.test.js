import React from 'react';
import { shallow, mount } from 'enzyme';
import App from '../App';
import beersMock from './__mocks__/beers.mock';

const mockResponse = (status, statusText, response) => new window.Response(response, {
    status,
    statusText,
    headers: {
        'Content-type': 'application/json',
    },
});

describe('App Unit Test', () => {
    let props;
    let shallowedApp;

    const getApp = () => {
        if (!shallowedApp) {
            shallowedApp = shallow(<App {...props} />);
        }
        return shallowedApp;
    };

    beforeEach(() => {
        props = {};
        shallowedApp = undefined;
    });

    test('Component should render without crashing', () => {
        expect(getApp()).toBeDefined();
    });

    test('Component should render correctly', () => {
        expect(getApp()).toMatchSnapshot();
    });

    test('Should mount and call componentDidMount and getBeers without crashing', () => {
        global.fetch = jest.fn().mockImplementation(
            () => Promise.resolve(mockResponse(200, null, JSON.stringify(beersMock))),
        );
        const componentDidMount = jest.spyOn(App.prototype, 'componentDidMount');
        const getBeersSpy = jest.spyOn(App.prototype, 'getBeers');
        const wrapper = mount(<App />);
        expect(wrapper).toBeDefined();
        expect(componentDidMount).toHaveBeenCalled();
        expect(getBeersSpy).toHaveBeenCalled();
    });

    test('Should call getBeers without crashing', () => {
        const wrapper = getApp();
        wrapper.instance().getBeers = jest.fn();
        wrapper.instance().onFilter();
        expect(wrapper.instance().getBeers).toHaveBeenCalled();
    });

    test('Should render Loader without crashing', () => {
        const wrapper = getApp();
        wrapper.setState({ isFetching: true });
        expect(wrapper.find('Loader')).toBeDefined();
    });

    test('Should render BeerGrid without crashing', () => {
        const wrapper = getApp();
        wrapper.setState({ isFetching: true });
        wrapper.setState({ data: beersMock });
        expect(wrapper.find('BeerGrid')).toBeDefined();
    });

    test('Should render Message without crashing', () => {
        const wrapper = getApp();
        wrapper.setState({ hasError: true });
        expect(wrapper.find('Message')).toBeDefined();
    });

    test('Should set correct state when API response OK', async () => {
        global.fetch = jest.fn().mockImplementation(
            () => Promise.resolve(mockResponse(200, null, JSON.stringify(beersMock))),
        );
        const wrapper = getApp();
        await wrapper.instance().getBeers({ testFilter: 'test' });
        expect(wrapper.state('hasError')).toEqual(false);
        expect(wrapper.state('isFetching')).toEqual(false);
        expect(wrapper.state('data')).toEqual(beersMock);
    });

    test('Should set correct state when API response Error', async () => {
        global.fetch = jest.fn().mockImplementation(
            () => Promise.resolve(mockResponse(400, null, JSON.stringify({ error: 'Bad Request' }))),
        );
        const wrapper = getApp();
        await wrapper.instance().getBeers();
        expect(wrapper.state('hasError')).toEqual(true);
        expect(wrapper.state('isFetching')).toEqual(false);
        expect(wrapper.state('data')).toEqual(null);
    });

    test('Should set correct state when API response cannot be parsed ', async () => {
        global.fetch = jest.fn().mockImplementation(
            () => Promise.resolve(mockResponse(400, null, 'Bad JSON')),
        );
        const wrapper = getApp();
        await wrapper.instance().getBeers();
        expect(wrapper.state('hasError')).toEqual(true);
        expect(wrapper.state('isFetching')).toEqual(false);
        expect(wrapper.state('data')).toEqual(null);
    });
});
