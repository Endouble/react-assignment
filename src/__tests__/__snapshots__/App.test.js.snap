// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Unit Test Component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="app-header"
        >
          <Header
            as="h1"
            inverted={true}
          >
            <Icon
              as="i"
              name="beer"
            />
            <HeaderContent>
              Punk Beers
              <HeaderSubheader>
                by Nicolas Burghi
              </HeaderSubheader>
            </HeaderContent>
          </Header>
        </header>,
        <Container>
          <Filter
            onFilter={[Function]}
          />
          <Loader
            active={true}
            inline="centered"
          />
        </Container>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Header
            as="h1"
            inverted={true}
          >
            <Icon
              as="i"
              name="beer"
            />
            <HeaderContent>
              Punk Beers
              <HeaderSubheader>
                by Nicolas Burghi
              </HeaderSubheader>
            </HeaderContent>
          </Header>,
          "className": "app-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "h1",
            "children": Array [
              <Icon
                as="i"
                name="beer"
              />,
              <HeaderContent>
                Punk Beers
                <HeaderSubheader>
                  by Nicolas Burghi
                </HeaderSubheader>
              </HeaderContent>,
            ],
            "inverted": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "i",
                "name": "beer",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  "Punk Beers",
                  <HeaderSubheader>
                    by Nicolas Burghi
                  </HeaderSubheader>,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Punk Beers",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": "by Nicolas Burghi",
                  },
                  "ref": null,
                  "rendered": "by Nicolas Burghi",
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Filter
              onFilter={[Function]}
            />,
            <Loader
              active={true}
              inline="centered"
            />,
            false,
            false,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onFilter": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "active": true,
              "inline": "centered",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          false,
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="app-header"
          >
            <Header
              as="h1"
              inverted={true}
            >
              <Icon
                as="i"
                name="beer"
              />
              <HeaderContent>
                Punk Beers
                <HeaderSubheader>
                  by Nicolas Burghi
                </HeaderSubheader>
              </HeaderContent>
            </Header>
          </header>,
          <Container>
            <Filter
              onFilter={[Function]}
            />
            <Loader
              active={true}
              inline="centered"
            />
          </Container>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Header
              as="h1"
              inverted={true}
            >
              <Icon
                as="i"
                name="beer"
              />
              <HeaderContent>
                Punk Beers
                <HeaderSubheader>
                  by Nicolas Burghi
                </HeaderSubheader>
              </HeaderContent>
            </Header>,
            "className": "app-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": "h1",
              "children": Array [
                <Icon
                  as="i"
                  name="beer"
                />,
                <HeaderContent>
                  Punk Beers
                  <HeaderSubheader>
                    by Nicolas Burghi
                  </HeaderSubheader>
                </HeaderContent>,
              ],
              "inverted": true,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "i",
                  "name": "beer",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": Array [
                    "Punk Beers",
                    <HeaderSubheader>
                      by Nicolas Burghi
                    </HeaderSubheader>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "Punk Beers",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "children": "by Nicolas Burghi",
                    },
                    "ref": null,
                    "rendered": "by Nicolas Burghi",
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Filter
                onFilter={[Function]}
              />,
              <Loader
                active={true}
                inline="centered"
              />,
              false,
              false,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "onFilter": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "active": true,
                "inline": "centered",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
            false,
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
